### DISCUSSION TITLE
mismatched types and web:Data<> problem when called from different file

### DISCUSSION THREAD

**Discussion by Leanerz** [actix/actix-web#2342] [Q&A] [answered]
Hi all,

I have rust folder like so:
```
main.rs
api/
- user_api/
-- mod.rs
-- routes.rs
-- user_api.rs
```

mod.rs:
```rust
mod routes;
mod user_api;

pub use routes::init_routes;
pub use user_api::{AddUsers, testOnly};
```

This function is inside: user_api.rs file:
```rust
    pub async fn AddUsers(cgrpc: APIClient<tonic::transport::Channel>, uid: u64) -> Result<TResponse<UserResponse>, Status> {
            let request = tonic::Request::new(UserUID {
                   uid
            });
        

            let response = cgrpc.clone().nuser(request).await;

            println!("RESPONSE={:?}", response);

            response
    }
```

This one inside routes.rs:
```rust
#[post("/adduser")]
async fn add_user(_req: HttpRequest, cgrpc: web::Data<APIClient<tonic::transport::Channel>>) -> HttpResponse {

    user_api::testOnly(1);

    match AddUsers(cgrpc, 1) {
        Ok(result) => HttpResponse::Ok().json(result.unwrap()),
        Err(err) => HttpResponse::Ok().json(err),
    }
}
```

But I get this error for cgrpc:

```
mismatched types

expected struct `api::user_api::user_api::client_grpc_api::gateway_client::APIClient`, found struct `actix_web::web::Data`

note: expected struct `api::user_api::user_api::client_grpc_api::gateway_client::APIClient<tonic::transport::Channel>`
         found struct `actix_web::web::Data<api::user_api::routes::client_grpc_api::gateway_client::APIClient<tonic::transport::Channel>>`rustc(E0308)
```

Inside main.rs:

```rust

    ...
    HttpServer::new(move || {
        App::new()
            .data(cgrpc.clone())
            .wrap(middleware::Compress::new(ContentEncoding::Br))
            .wrap(middleware::Logger::default())
            .service(web::scope("/user").configure(api::user_api::init_routes))
    })
    .bind("127.0.0.1:80")?
    .run()
    .await
```

I can copy paste all functions inside user_api.rs to routes.rs and use directly:

```rust
web::Data<APIClient<tonic::transport::Channel>>
```

But for organization purpose, I need it like so, your help will be appreciated.

Thank you !

**Comment by aliemjay** [comment, marked as answer]
In `add_user`, `cgrpc` is of type `Data`. You can only get `&APIClient<_>` from it by either using `Data::get_ref` or by Deref trait and deref coercion. However, you should modify `AddUsers` to receive a  `&APIClient`, instead of `APIClient` (or you can clone it!).

Here is an example.

Modify AddUsers:
```rust
pub async fn AddUsers(cgrpc: &APIClient<tonic::transport::Channel> ...
```

in add_user (you also need to await):
```rust
AddUsers(cgrpc.get_ref(), 1).await;
//  or by deref coercion
AddUsers(&cgrpc, 1).await;
```

**Reply by Leanerz** [reply]
Thank you for your answer !

This solve the problem :) another question please, you said that I have to use await, what is the correct way for that please?

Let say I want to send a GRPC request upon recieving new POST request from REST API, and since it's an async function and no need to wait for a response but only the status, how I can manage to do it?

After some searchs, I found about `actix::Arbiter`, is it the right way? If yes, how to combine it with AddUsers() function for example?

Sorry again for taking from your time :)

**Reply by aliemjay** [reply]
> another question please, you said that I have to use await, what is the correct way for that please?

Just a `.await` :) I've modified the answer.

I don't use tonic, but by taking a quick look at it, I can say you are on the right track. You construct the client globally and pass it as `Data<_>` to use it in the handlers. No need to use `actix::Arbiter` though.

**Reply by aliemjay** [reply]
You also need to constuct the client once per each actix worker thread.

Something like this (assuming you consreate the channel in APICLient::new)

```rust
    HttpServer::new(move || {
        let cgrpc = APIClient::new();

        App::new()
            .data(cgrpc)
            .wrap(middleware::Compress::new(ContentEncoding::Br))
            .wrap(middleware::Logger::default())
            .service(web::scope("/user").configure(api::user_api::init_routes))
    })
```

**Reply by Leanerz** [reply]
I did the same thing (built the grpc client - APIClient::new) but before:

```rust
HttpServer::new(move || {
```

Result is something like this:

```rust
    let endpoints = ["http://[::1]:50051", "http://[::1]:50052"].iter().map(|a| Channel::from_static(a));
    let channel = Channel::balance_list(endpoints);
    let cgrpc = APIClient::new(channel);

    HttpServer::new(move || {
        App::new()
            .data(cgrpc.clone())
            .wrap(middleware::Compress::new(ContentEncoding::Br))
            .wrap(middleware::Logger::default())
            .service(web::scope("/user").configure(api::user_api::init_routes))
    })
```

It should be inside it? I use regulary for example MySQL pool and Redis outside it also.

Should I move them inside it?

Thanks again for all this useful informations you are sharing ! sorry for annoying you again !

**Reply by aliemjay** [reply]
> It should be inside it?

I see the `Channel` is  Send+Sync. So the answer is no, I guess. If it compiles then it should be ok. Do you have problems with it?

and you're welcome

**Reply by Leanerz** [reply]
It compile without a problem when it is before `HttpServer::new(move || {`, but if it is preferable to put it inside I will do that.

Wondering what is the difference.

🙏 🙏 🙏

**Reply by robjtede** [reply]
> Wondering what is the difference.

https://actix.rs/actix-web/actix_web/struct.App.html#shared-mutable-state

**Reply by Leanerz** [reply]
> 
> 
> > Wondering what is the difference.
> 
> https://actix.rs/actix-web/actix_web/struct.App.html#shared-mutable-state

In the example you shared it looks like as I've always did, but was curious to know what is the difference if I put the build of grpc client inside `HttpServer::new(move || {`.

Just by guess, I think in the second case, I don't have to clone the client like in the official example since it's inside and he execute it for each new request (Which is same at the end).

Thank you for giving me the opportunity to learn more about it.

I appreciate it :)

**Reply by robjtede** [reply]
Put simply:
Created with Arc/Data outside HttpServer: 1 instance
Created inside HttpServer with N workers: N instances

**Reply by Leanerz** [reply]
> 
> 
> Put simply:
> Created with Arc/Data outside HttpServer: 1 instance
> Created inside HttpServer with N workers: N instances

Got it now ! Thank you !
